ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java

```diff
@@ -83,10 +85,12 @@ public abstract class RaftLogBase implements RaftLog {
   private final long purgePreservation;
 
   private final AtomicReference<LogEntryProto> lastMetadataEntry = new AtomicReference<>();
+  private final Function<StateMachineLogEntryProto, String> stateMachineToString;
```

Comment:
Let don't add `StateMachine` dependency `RaftLogBase` . The subclasses can override `toLogEntryString(..)`.